WMIC missing, breaks installers on newer Windows versions

Windows 24H2 and higher.

```
[INFO] 2025-04-11T19:28:10,015 @ qz.installer.Installer:94
        Stopping running instances...
[DEBUG] 2025-04-11T19:28:10,149 @ qz.utils.ShellUtilities:179
        Executing: [wmic.exe, process, where, (Name='java.exe' OR Name='javaw.exe'), get, processid]
[ERROR] 2025-04-11T19:28:10,161 @ qz.utils.ShellUtilities:199
        IOException executing: [wmic.exe, process, where, (Name='java.exe' OR Name='javaw.exe'), get, processid] envp: (suppressed)
java.io.IOException: Cannot run program "wmic.exe": CreateProcess error=2, The system cannot find the file specified
        at java.base/java.lang.ProcessBuilder.start(Unknown Source) ~[?:?]
        at java.base/java.lang.ProcessBuilder.start(Unknown Source) ~[?:?]
        at java.base/java.lang.Runtime.exec(Unknown Source) ~[?:?]
        at java.base/java.lang.Runtime.exec(Unknown Source) ~[?:?]
        at qz.utils.ShellUtilities.executeRaw(ShellUtilities.java:184) [qz-tray.jar:?]
        at qz.utils.ShellUtilities.executeRaw(ShellUtilities.java:168) [qz-tray.jar:?]
        at qz.installer.certificate.firefox.locator.WindowsAppLocator.getPids(WindowsAppLocator.java:69) [qz-tray.jar:?]
        at qz.installer.certificate.firefox.locator.AppLocator.getPids(AppLocator.java:22) [qz-tray.jar:?]
        at qz.installer.TaskKiller.killAll(TaskKiller.java:52) [qz-tray.jar:?]
        at qz.installer.Installer.preinstall(Installer.java:95) [qz-tray.jar:?]
        at qz.utils.ArgParser.processInstallerArgs(ArgParser.java:178) [qz-tray.jar:?]
        at qz.utils.ArgParser.intercept(ArgParser.java:360) [qz-tray.jar:?]
        at qz.App.main(App.java:37) [qz-tray.jar:?]
```

The workaround is to enable the optional Windows Feature called "WMIC".  The long-term fix will be more work.

![Image](https://github.com/user-attachments/assets/f3f8a651-62ac-485a-b635-13f28fa60897)

--- a/ant/apple/installer.xml
+++ b/ant/apple/installer.xml
@@ -258,6 +258,7 @@
             <arg value="--entitlement"/>
             <arg value="${build.dir}/apple-entitlements.plist"/>
             <arg value="${bundle.dir}/Contents/PlugIns/Java.runtime/Contents/Home/bin/java"/>
+            <arg value="${bundle.dir}/Contents/PlugIns/Java.runtime/Contents/Home/bin/jcmd"/>
             <arg value="${bundle.dir}/Contents/PlugIns/Java.runtime"/>
         </exec>
         <exec executable="codesign" failonerror="true">

--- a/src/qz/build/JLink.java
+++ b/src/qz/build/JLink.java
@@ -236,6 +236,8 @@ public class JLink {
                 // Java accessibility bridge dependency, see https://github.com/qzind/tray/issues/1234
                 depList.add("jdk.accessibility");
             default:
+                // Adds "bin/jcmd"
+                depList.add("jdk.jcmd");
                 // "jar:" URLs create transient zipfs dependency, see https://stackoverflow.com/a/57846672/3196753
                 depList.add("jdk.zipfs");
                 // fix for https://github.com/qzind/tray/issues/894 solution from https://github.com/adoptium/adoptium-support/issues/397

--- a/src/qz/build/JLink.java
+++ b/src/qz/build/JLink.java
@@ -294,6 +294,7 @@ public class JLink {
                 case WINDOWS:
                     keepFiles.add("java.exe");
                     keepFiles.add("javaw.exe");
+                    keepFiles.add("jcmd.exe");
                     if(depList.contains("jdk.accessibility")) {
                         // Java accessibility bridge switching tool
                         keepFiles.add("jabswitch.exe");
@@ -303,6 +304,7 @@ public class JLink {
                     break;
                 default:
                     keepFiles.add("java");
+                    keepFiles.add("jcmd");
                     keepExt = null;
             }
 

--- a/src/qz/installer/TaskKiller.java
+++ b/src/qz/installer/TaskKiller.java
@@ -12,11 +12,14 @@ package qz.installer;
 import org.apache.commons.lang3.StringUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import qz.common.Constants;
 import qz.installer.certificate.firefox.locator.AppLocator;
 import qz.utils.ShellUtilities;
 import qz.utils.SystemUtilities;
 import qz.ws.PrintSocketServer;
 
+import java.io.IOException;
+import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
@@ -25,74 +28,90 @@ import static qz.common.Constants.PROPS_FILE;
 
 public class TaskKiller {
     protected static final Logger log = LogManager.getLogger(TaskKiller.class);
-    private static final String[] TRAY_PID_QUERY_POSIX = {"pgrep", "-f", PROPS_FILE + ".jar" };
-    private static final String[] KILL_PID_CMD_POSIX = {"kill", "-9", ""/*pid placeholder*/};
-
-    private static final String[] TRAY_PID_QUERY_WIN32 = {"wmic.exe", "process", "where", "CommandLine like '%" + PROPS_FILE + ".jar" + "%'", "get", "processid" };
-    private static final String[] KILL_PID_CMD_WIN32 = {"taskkill.exe", "/F", "/PID", "" /*pid placeholder*/ };
+    private static final String[] JAR_NAMES = { PROPS_FILE + ".jar" };
+    private static final String[] KILL_PID_CMD_POSIX = { "kill", "-9" };
+    private static final String[] KILL_PID_CMD_WIN32 = { "taskkill.exe", "/F", "/PID" };
+    private static final String[] KILL_PID_CMD = SystemUtilities.isWindows() ? KILL_PID_CMD_WIN32 : KILL_PID_CMD_POSIX;
 
     /**
      * Kills all QZ Tray processes, being careful not to kill itself
      */
     public static boolean killAll() {
-        // TODO:
-        // 1. Calculate the path to jcmd
-        // 2. List all Java processes using jcmd -l
-        // 3. Filter Java processes matching "qz-tray.jar"
-        // 4. Filter Java processes matching "PrintSocketClient" (IntelliJ)
-        // 5. Kill all matching processes
-
         boolean success = true;
 
-        ArrayList<String> javaProcs;
-        String[] trayProcs;
-        int selfProc = SystemUtilities.getProcessId();
-        String[] killCmd;
         // Disable service until reboot
         if(SystemUtilities.isMac()) {
             ShellUtilities.execute("/bin/launchctl", "unload", MacInstaller.LAUNCH_AGENT_PATH);
         }
-        if(SystemUtilities.isWindows()) {
-            // Windows may be running under javaw.exe (normal) or java.exe (terminal)
-            javaProcs = AppLocator.getInstance().getPids("java.exe", "javaw.exe");
-            trayProcs = ShellUtilities.executeRaw(TRAY_PID_QUERY_WIN32).split("\\s*\\r?\\n");
-            killCmd = KILL_PID_CMD_WIN32;
-        } else {
-            javaProcs = AppLocator.getInstance().getPids( "java");
-            trayProcs = ShellUtilities.executeRaw(TRAY_PID_QUERY_POSIX).split("\\s*\\r?\\n");
-            killCmd = KILL_PID_CMD_POSIX;
-        }
-        if (!javaProcs.isEmpty()) {
-            // Find intersections of java and qz-tray.jar
-            List<String> intersections = new ArrayList<>(Arrays.asList(trayProcs));
-            intersections.retainAll(javaProcs);
 
-            // Remove any instances created by this installer
-            intersections.remove("" + selfProc);
-
-            // Kill whatever's left
-            for (String pid : intersections) {
-                // isNumeric() needed for Windows; filters whitespace, headers
-                if(StringUtils.isNumeric(pid)) {
-                    // Set last command to the pid
-                    killCmd[killCmd.length -1] = pid;
-                    success = success && ShellUtilities.execute(killCmd);
-                }
-            }
+        // Get the matching PIDs
+        ArrayList<Integer> pids;
+        try {
+            pids = findTrayPids();
+        } catch(IOException e) {
+            log.error("Failed to retrieve PIDs for {}", Constants.ABOUT_TITLE, e);
+            return false;
         }
 
-        // Use jcmd to kill class processes too, such as through the IDE
-        if(SystemUtilities.isJDK()) {
-            String[] procs = ShellUtilities.executeRaw("jcmd", "-l").split("\\r?\\n");
-            for(String proc : procs) {
-                String[] parts = proc.split(" ", 1);
-                if (parts.length >= 2 && parts[1].contains(PrintSocketServer.class.getCanonicalName())) {
-                    killCmd[killCmd.length - 1] = parts[0].trim();
-                    success = success && ShellUtilities.execute(killCmd);
-                }
-            }
+        // Kill each PID
+        String[] killPid = new String[KILL_PID_CMD.length + 1];
+        for (Integer pid : pids) {
+            killPid[killPid.length - 1] = pid.toString();
+            success = success && ShellUtilities.execute(killPid);
         }
 
         return success;
+    }
+
+    private static Path getJcmdPath() throws IOException {
+        Path jcmd;
+        if(SystemUtilities.isWindows()) {
+            jcmd = SystemUtilities.getJarParentPath().resolve("/runtime/bin/jcmd.exe");
+        } else if (SystemUtilities.isMac()) {
+            jcmd = SystemUtilities.getJarParentPath().resolve("../PlugIns/Java.runtime/Contents/Home/bin/jcmd");
+        } else {
+            jcmd = SystemUtilities.getJarParentPath().resolve("../PlugIns/Java.runtime/Contents/Home/bin/jcmd");
+        }
+        if(!jcmd.toFile().exists()) {
+            throw new IOException("Could not find jcmd, we can't stop running instances");
+        }
+        return jcmd;
+    }
+
+
+    /**
+     * Uses jcmd to fetch all PIDs that match this product
+     */
+    private static ArrayList<Integer> findTrayPids() throws IOException {
+        ArrayList<Integer> foundProcs = new ArrayList<>();
+
+        String[] stdout = ShellUtilities.executeRaw(getJcmdPath().toString(), "-l").split("\\r?\\n");
+        if(stdout == null || stdout.length == 0) {
+            throw new IOException("Error calling '" + getJcmdPath() + "' -l");
+        }
+        for(String line : stdout) {
+            // e.g. "35446 C:\Program Files\QZ Tray\qz-tray.jar"
+            String[] parts = line.split(" ", 1);
+            if (parts.length >= 2) {
+                Integer proc = Integer.parseInt(parts[0]);
+                String command = parts[1];
+                // Handle running from IDE such as IntelliJ
+                if(command.contains(PrintSocketServer.class.getCanonicalName())) {
+                    foundProcs.add(proc);
+                    continue;
+                }
+                // Handle "qz-tray.jar"
+                for(String jarName : JAR_NAMES) {
+                    if(command.contains(jarName));
+                    foundProcs.add(proc);
+                    break; // continue parent loop
+                }
+            }
+        }
+
+        // Careful not to kill ourselves ;)
+        foundProcs.remove(SystemUtilities.getProcessId());
+
+        return foundProcs;
     }
 }

--- a/src/qz/installer/TaskKiller.java
+++ b/src/qz/installer/TaskKiller.java
@@ -46,7 +46,7 @@ public class TaskKiller {
         // Use jcmd to get all java processes
         HashSet<Integer> pids = findPidsJcmd();
         if(!SystemUtilities.isWindows()) {
-            // Fallback to pgrep (See liberica/QZ-23)
+            // Fallback to pgrep, needed for macOS (See JDK-8319589, JDK-8197387)
             pids.addAll(findPidsPgrep());
         }
 

--- a/src/qz/installer/TaskKiller.java
+++ b/src/qz/installer/TaskKiller.java
@@ -13,7 +13,6 @@ import org.apache.commons.lang3.StringUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import qz.common.Constants;
-import qz.installer.certificate.firefox.locator.AppLocator;
 import qz.utils.ShellUtilities;
 import qz.utils.SystemUtilities;
 import qz.ws.PrintSocketServer;
@@ -22,7 +21,7 @@ import java.io.IOException;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.List;
+import java.util.HashSet;
 
 import static qz.common.Constants.PROPS_FILE;
 
@@ -44,17 +43,16 @@ public class TaskKiller {
             ShellUtilities.execute("/bin/launchctl", "unload", MacInstaller.LAUNCH_AGENT_PATH);
         }
 
-        // Get the matching PIDs
-        ArrayList<Integer> pids;
-        try {
-            pids = findTrayPids();
-        } catch(IOException e) {
-            log.error("Failed to retrieve PIDs for {}", Constants.ABOUT_TITLE, e);
-            return false;
+        // Use jcmd to get all java processes
+        HashSet<Integer> pids = findPidsJcmd();
+        if(!SystemUtilities.isWindows()) {
+            // Fallback to pgrep (See liberica/QZ-23)
+            pids.addAll(findPidsPgrep());
         }
 
         // Kill each PID
         String[] killPid = new String[KILL_PID_CMD.length + 1];
+        System.arraycopy(KILL_PID_CMD, 0, killPid, 0, KILL_PID_CMD.length);
         for (Integer pid : pids) {
             killPid[killPid.length - 1] = pid.toString();
             log.debug("Found {} running as PID {}, calling '{}'", Constants.ABOUT_TITLE, pid, Arrays.toString(killPid));
@@ -71,48 +69,73 @@ public class TaskKiller {
         } else if (SystemUtilities.isMac()) {
             jcmd = SystemUtilities.getJarParentPath().resolve("../PlugIns/Java.runtime/Contents/Home/bin/jcmd");
         } else {
-            jcmd = SystemUtilities.getJarParentPath().resolve("../PlugIns/Java.runtime/Contents/Home/bin/jcmd");
+            jcmd = SystemUtilities.getJarParentPath().resolve("/runtime/bin/jcmd");
         }
         if(!jcmd.toFile().exists()) {
-            throw new IOException("Could not find jcmd, we can't stop running instances");
+            throw new IOException("Could not find jcmd, we can't use it for detecting running instances");
         }
         return jcmd;
+    }
+
+    private static HashSet<Integer> findPidsPgrep() {
+        HashSet<Integer> foundPids = new HashSet<>();
+
+        for(String jarName : JAR_NAMES) {
+            String[] pids = ShellUtilities.executeRaw("pgrep", "-f", jarName).split("\\s*\\r?\\n");
+            for(String pid : pids) {
+                if (StringUtils.isNumeric(pid)) {
+                    foundPids.add(Integer.parseInt(pid));
+                }
+            }
+        }
+
+        // Careful not to kill ourselves ;)
+        foundPids.remove(SystemUtilities.getProcessId());
+
+        return foundPids;
     }
 
 
     /**
      * Uses jcmd to fetch all PIDs that match this product
      */
-    private static ArrayList<Integer> findTrayPids() throws IOException {
-        ArrayList<Integer> foundProcs = new ArrayList<>();
+    private static HashSet<Integer> findPidsJcmd() {
+        HashSet<Integer> foundPids = new HashSet<>();
 
-        String[] stdout = ShellUtilities.executeRaw(getJcmdPath().toString(), "-l").split("\\r?\\n");
-        if(stdout == null || stdout.length == 0) {
-            throw new IOException("Error calling '" + getJcmdPath() + "' -l");
+        String[] stdout = {};
+        try {
+            stdout = ShellUtilities.executeRaw(getJcmdPath().toString(), "-l").split("\\r?\\n");
+            if(stdout == null || stdout.length == 0) {
+               log.error("Error calling '{}' {}", getJcmdPath(), "-l");
+            }
+        } catch(Exception e) {
+            log.error(e);
+            return foundPids;
         }
+
         for(String line : stdout) {
             // e.g. "35446 C:\Program Files\QZ Tray\qz-tray.jar"
             String[] parts = line.split(" ", 1);
-            if (parts.length >= 2) {
+            if (parts.length >= 2 && StringUtils.isNumeric(parts[0])) {
                 Integer proc = Integer.parseInt(parts[0]);
                 String command = parts[1];
                 // Handle running from IDE such as IntelliJ
                 if(command.contains(PrintSocketServer.class.getCanonicalName())) {
-                    foundProcs.add(proc);
+                    foundPids.add(proc);
                     continue;
                 }
                 // Handle "qz-tray.jar"
                 for(String jarName : JAR_NAMES) {
                     if(command.contains(jarName));
-                    foundProcs.add(proc);
+                    foundPids.add(proc);
                     break; // continue parent loop
                 }
             }
         }
 
         // Careful not to kill ourselves ;)
-        foundProcs.remove(SystemUtilities.getProcessId());
+        foundPids.remove(SystemUtilities.getProcessId());
 
-        return foundProcs;
+        return foundPids;
     }
 }

--- a/src/qz/installer/TaskKiller.java
+++ b/src/qz/installer/TaskKiller.java
@@ -71,7 +71,7 @@ public class TaskKiller {
             jcmd = SystemUtilities.getJarParentPath().resolve("runtime/bin/jcmd");
         }
         if(!jcmd.toFile().exists()) {
-            log.error("Could not find find {}", jcmd);
+            log.error("Could not find {}", jcmd);
             throw new IOException("Could not find jcmd, we can't use it for detecting running instances");
         }
         return jcmd;

--- a/src/qz/installer/TaskKiller.java
+++ b/src/qz/installer/TaskKiller.java
@@ -65,13 +65,14 @@ public class TaskKiller {
     private static Path getJcmdPath() throws IOException {
         Path jcmd;
         if(SystemUtilities.isWindows()) {
-            jcmd = SystemUtilities.getJarParentPath().resolve("/runtime/bin/jcmd.exe");
+            jcmd = SystemUtilities.getJarParentPath().resolve("runtime/bin/jcmd.exe");
         } else if (SystemUtilities.isMac()) {
             jcmd = SystemUtilities.getJarParentPath().resolve("../PlugIns/Java.runtime/Contents/Home/bin/jcmd");
         } else {
-            jcmd = SystemUtilities.getJarParentPath().resolve("/runtime/bin/jcmd");
+            jcmd = SystemUtilities.getJarParentPath().resolve("runtime/bin/jcmd");
         }
         if(!jcmd.toFile().exists()) {
+            log.error("Could not find find {}", jcmd);
             throw new IOException("Could not find jcmd, we can't use it for detecting running instances");
         }
         return jcmd;

--- a/src/qz/installer/TaskKiller.java
+++ b/src/qz/installer/TaskKiller.java
@@ -12,22 +12,22 @@ package qz.installer;
 import org.apache.commons.lang3.StringUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
-import qz.common.Constants;
 import qz.utils.ShellUtilities;
 import qz.utils.SystemUtilities;
-import qz.ws.PrintSocketServer;
 
 import java.io.IOException;
 import java.nio.file.Path;
-import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.HashSet;
 
 import static qz.common.Constants.PROPS_FILE;
 
 public class TaskKiller {
     protected static final Logger log = LogManager.getLogger(TaskKiller.class);
-    private static final String[] JAR_NAMES = { PROPS_FILE + ".jar" };
+    private static final String[] JAR_NAMES = {
+            PROPS_FILE + ".jar",
+            "qz.App", // v2.2.0...
+            "qz.ws.PrintSocketServer" // v2.0.0...v2.1.6
+    };
     private static final String[] KILL_PID_CMD_POSIX = { "kill", "-9" };
     private static final String[] KILL_PID_CMD_WIN32 = { "taskkill.exe", "/F", "/PID" };
     private static final String[] KILL_PID_CMD = SystemUtilities.isWindows() ? KILL_PID_CMD_WIN32 : KILL_PID_CMD_POSIX;
@@ -55,7 +55,6 @@ public class TaskKiller {
         System.arraycopy(KILL_PID_CMD, 0, killPid, 0, KILL_PID_CMD.length);
         for (Integer pid : pids) {
             killPid[killPid.length - 1] = pid.toString();
-            log.debug("Found {} running as PID {}, calling '{}'", Constants.ABOUT_TITLE, pid, Arrays.toString(killPid));
             success = success && ShellUtilities.execute(killPid);
         }
 
@@ -84,9 +83,11 @@ public class TaskKiller {
         for(String jarName : JAR_NAMES) {
             String[] pids = ShellUtilities.executeRaw("pgrep", "-f", jarName).split("\\s*\\r?\\n");
             for(String pid : pids) {
-                if (StringUtils.isNumeric(pid)) {
-                    foundPids.add(Integer.parseInt(pid));
+                if (!StringUtils.isNumeric(pid.trim())) {
+                    log.warn("Found PID value '{}' that does not appear to be a number", pid);
+                    continue;
                 }
+                foundPids.add(Integer.parseInt(pid));
             }
         }
 
@@ -103,7 +104,7 @@ public class TaskKiller {
     private static HashSet<Integer> findPidsJcmd() {
         HashSet<Integer> foundPids = new HashSet<>();
 
-        String[] stdout = {};
+        String[] stdout;
         try {
             stdout = ShellUtilities.executeRaw(getJcmdPath().toString(), "-l").split("\\r?\\n");
             if(stdout == null || stdout.length == 0) {
@@ -116,21 +117,24 @@ public class TaskKiller {
 
         for(String line : stdout) {
             // e.g. "35446 C:\Program Files\QZ Tray\qz-tray.jar"
-            String[] parts = line.split(" ", 1);
-            if (parts.length >= 2 && StringUtils.isNumeric(parts[0])) {
-                Integer proc = Integer.parseInt(parts[0]);
-                String command = parts[1];
-                // Handle running from IDE such as IntelliJ
-                if(command.contains(PrintSocketServer.class.getCanonicalName())) {
-                    foundPids.add(proc);
+            String[] parts = line.split(" ", 2);
+            if (parts.length == 2) {
+                String pidString = parts[0].trim();
+                if(!StringUtils.isNumeric(pidString)) {
+                    log.warn("Found PID value '{}' that does not appear to be a number", pidString);
                     continue;
                 }
-                // Handle "qz-tray.jar"
+                Integer pid = Integer.parseInt(pidString);
+                String args = parts[1];
+
                 for(String jarName : JAR_NAMES) {
-                    if(command.contains(jarName));
-                    foundPids.add(proc);
-                    break; // continue parent loop
+                    if(args.contains(jarName)) {
+                        foundPids.add(pid);
+                        break; // continue parent loop
+                    }
                 }
+            } else {
+                log.warn("Found erroneous output, skipping", line);
             }
         }
 

--- a/src/qz/installer/TaskKiller.java
+++ b/src/qz/installer/TaskKiller.java
@@ -83,7 +83,12 @@ public class TaskKiller {
         for(String jarName : JAR_NAMES) {
             String[] pids = ShellUtilities.executeRaw("pgrep", "-f", jarName).split("\\s*\\r?\\n");
             for(String pid : pids) {
-                if (!StringUtils.isNumeric(pid.trim())) {
+                pid = pid.trim();
+                if(pid.isEmpty()) {
+                    // Don't try to process blank lines
+                    continue;
+                }
+                if (!StringUtils.isNumeric(pid)) {
                     log.warn("Found PID value '{}' that does not appear to be a number", pid);
                     continue;
                 }
@@ -104,18 +109,24 @@ public class TaskKiller {
     private static HashSet<Integer> findPidsJcmd() {
         HashSet<Integer> foundPids = new HashSet<>();
 
-        String[] stdout;
+        String[] lines;
         try {
-            stdout = ShellUtilities.executeRaw(getJcmdPath().toString(), "-l").split("\\r?\\n");
-            if(stdout == null || stdout.length == 0) {
+            String stdout = ShellUtilities.executeRaw(getJcmdPath().toString(), "-l");
+            if(stdout == null) {
                log.error("Error calling '{}' {}", getJcmdPath(), "-l");
+               return foundPids;
             }
+            lines = stdout.split("\\r?\\n");
         } catch(Exception e) {
             log.error(e);
             return foundPids;
         }
 
-        for(String line : stdout) {
+        for(String line : lines) {
+            if(line.trim().isEmpty()) {
+                // Don't try to process blank lines
+                continue;
+            }
             // e.g. "35446 C:\Program Files\QZ Tray\qz-tray.jar"
             String[] parts = line.split(" ", 2);
             if (parts.length == 2) {
@@ -134,7 +145,7 @@ public class TaskKiller {
                     }
                 }
             } else {
-                log.warn("Found erroneous output, skipping", line);
+                log.warn("Found erroneous output: '{}', skipping", line);
             }
         }
 

--- a/src/qz/installer/TaskKiller.java
+++ b/src/qz/installer/TaskKiller.java
@@ -15,10 +15,8 @@ import org.apache.logging.log4j.Logger;
 import qz.installer.certificate.firefox.locator.AppLocator;
 import qz.utils.ShellUtilities;
 import qz.utils.SystemUtilities;
-import qz.utils.WindowsUtilities;
 import qz.ws.PrintSocketServer;
 
-import java.io.File;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
@@ -37,6 +35,13 @@ public class TaskKiller {
      * Kills all QZ Tray processes, being careful not to kill itself
      */
     public static boolean killAll() {
+        // TODO:
+        // 1. Calculate the path to jcmd
+        // 2. List all Java processes using jcmd -l
+        // 3. Filter Java processes matching "qz-tray.jar"
+        // 4. Filter Java processes matching "PrintSocketClient" (IntelliJ)
+        // 5. Kill all matching processes
+
         boolean success = true;
 
         ArrayList<String> javaProcs;
@@ -85,13 +90,6 @@ public class TaskKiller {
                     killCmd[killCmd.length - 1] = parts[0].trim();
                     success = success && ShellUtilities.execute(killCmd);
                 }
-            }
-        }
-
-        if(WindowsUtilities.isWindowsXP()) {
-            File f = new File("TempWmicBatchFile.bat");
-            if(f.exists()) {
-                f.deleteOnExit();
             }
         }
 

--- a/src/qz/installer/TaskKiller.java
+++ b/src/qz/installer/TaskKiller.java
@@ -50,6 +50,9 @@ public class TaskKiller {
             pids.addAll(findPidsPgrep());
         }
 
+        // Careful not to kill ourselves ;)
+        pids.remove(SystemUtilities.getProcessId());
+
         // Kill each PID
         String[] killPid = new String[KILL_PID_CMD.length + 1];
         System.arraycopy(KILL_PID_CMD, 0, killPid, 0, KILL_PID_CMD.length);
@@ -81,23 +84,22 @@ public class TaskKiller {
         HashSet<Integer> foundPids = new HashSet<>();
 
         for(String jarName : JAR_NAMES) {
-            String[] pids = ShellUtilities.executeRaw("pgrep", "-f", jarName).split("\\s*\\r?\\n");
-            for(String pid : pids) {
-                pid = pid.trim();
-                if(pid.isEmpty()) {
+            String stdout = ShellUtilities.executeRaw("pgrep", "-f", jarName);
+            String[] lines = stdout.split("\\s*\\r?\\n");
+            for(String line : lines) {
+                if(line.trim().isEmpty()) {
                     // Don't try to process blank lines
                     continue;
                 }
-                if (!StringUtils.isNumeric(pid)) {
-                    log.warn("Found PID value '{}' that does not appear to be a number", pid);
-                    continue;
+
+                int pid = parsePid(line);
+                if (pid >= 0) {
+                    foundPids.add(pid);
+                } else {
+                    log.warn("Could not parse PID value.  Full line: '{}', Full output: '{}'", line, stdout);
                 }
-                foundPids.add(Integer.parseInt(pid));
             }
         }
-
-        // Careful not to kill ourselves ;)
-        foundPids.remove(SystemUtilities.getProcessId());
 
         return foundPids;
     }
@@ -109,9 +111,10 @@ public class TaskKiller {
     private static HashSet<Integer> findPidsJcmd() {
         HashSet<Integer> foundPids = new HashSet<>();
 
+        String stdout;
         String[] lines;
         try {
-            String stdout = ShellUtilities.executeRaw(getJcmdPath().toString(), "-l");
+            stdout = ShellUtilities.executeRaw(getJcmdPath().toString(), "-l");
             if(stdout == null) {
                log.error("Error calling '{}' {}", getJcmdPath(), "-l");
                return foundPids;
@@ -123,35 +126,59 @@ public class TaskKiller {
         }
 
         for(String line : lines) {
-            if(line.trim().isEmpty()) {
+            if (line.trim().isEmpty()) {
                 // Don't try to process blank lines
                 continue;
             }
             // e.g. "35446 C:\Program Files\QZ Tray\qz-tray.jar"
             String[] parts = line.split(" ", 2);
-            if (parts.length == 2) {
-                String pidString = parts[0].trim();
-                if(!StringUtils.isNumeric(pidString)) {
-                    log.warn("Found PID value '{}' that does not appear to be a number", pidString);
+            int pid = parsePid(parts);
+            if (pid >= 0) {
+                String args = parseArgs(parts);
+                if (args == null) {
+                    log.warn("Found PID value '{}' but no args to match.  Full line: '{}', Full output: '{}'", pid, line, stdout);
                     continue;
                 }
-                Integer pid = Integer.parseInt(pidString);
-                String args = parts[1];
-
                 for(String jarName : JAR_NAMES) {
-                    if(args.contains(jarName)) {
+                    if (args.contains(jarName)) {
                         foundPids.add(pid);
                         break; // continue parent loop
                     }
                 }
             } else {
-                log.warn("Found erroneous output: '{}', skipping", line);
+                log.warn("Could not parse PID value.  Full line: '{}', Full output: '{}'", line, stdout);
             }
         }
 
-        // Careful not to kill ourselves ;)
-        foundPids.remove(SystemUtilities.getProcessId());
-
         return foundPids;
+    }
+
+    // Returns the second index of a String[], trimmed
+    private static String parseArgs(String[] input) {
+        if(input != null) {
+            if(input.length == 2) {
+                return input[1].trim();
+            }
+        }
+        return null;
+    }
+
+    // Parses an int value form the first index of a String[], returning -1 if something went wrong
+    private static int parsePid(String[] input) {
+        if(input != null) {
+            if(input.length == 2) {
+                return parsePid(input[0]);
+            }
+        }
+        return -1;
+    }
+
+    // Parses an int value form the provided string, returning -1 if something went wrong
+    private static int parsePid(String input) {
+        String pidString = input.trim();
+        if(StringUtils.isNumeric(pidString)) {
+            return Integer.parseInt(pidString);
+        }
+        return -1;
     }
 }

--- a/src/qz/installer/TaskKiller.java
+++ b/src/qz/installer/TaskKiller.java
@@ -57,6 +57,7 @@ public class TaskKiller {
         String[] killPid = new String[KILL_PID_CMD.length + 1];
         for (Integer pid : pids) {
             killPid[killPid.length - 1] = pid.toString();
+            log.debug("Found {} running as PID {}, calling '{}'", Constants.ABOUT_TITLE, pid, Arrays.toString(killPid));
             success = success && ShellUtilities.execute(killPid);
         }
 

--- a/src/qz/installer/certificate/firefox/locator/AppLocator.java
+++ b/src/qz/installer/certificate/firefox/locator/AppLocator.java
@@ -18,6 +18,7 @@ public abstract class AppLocator {
     public abstract ArrayList<AppInfo> locate(AppAlias appAlias);
     public abstract ArrayList<Path> getPidPaths(ArrayList<String> pids);
 
+    @SuppressWarnings("unused")
     public ArrayList<String> getPids(String ... processNames) {
         return getPids(new ArrayList<>(Arrays.asList(processNames)));
     }

--- a/src/qz/installer/certificate/firefox/locator/WindowsAppLocator.java
+++ b/src/qz/installer/certificate/firefox/locator/WindowsAppLocator.java
@@ -58,7 +58,10 @@ public class WindowsAppLocator extends AppLocator{
 
     @Override
     public ArrayList<String> getPids(ArrayList<String> processNames) {
-        ArrayList<String> javaPids = new ArrayList<>();
+        ArrayList<String> pidList = new ArrayList<>();
+
+        if (processNames.isEmpty()) return pidList;
+
         Tlhelp32.PROCESSENTRY32 pe32 = new Tlhelp32.PROCESSENTRY32();
         pe32.dwSize = new WinNT.DWORD(pe32.size());
 
@@ -66,26 +69,26 @@ public class WindowsAppLocator extends AppLocator{
         WinNT.HANDLE hSnapshot = Kernel32.INSTANCE.CreateToolhelp32Snapshot(Tlhelp32.TH32CS_SNAPPROCESS, new WinNT.DWORD(0));
         if (hSnapshot.equals(WinNT.INVALID_HANDLE_VALUE)) {
             log.warn("Process snapshot has invalid handle");
-            return javaPids;
+            return pidList;
         }
 
         if (Kernel32.INSTANCE.Process32First(hSnapshot, pe32)) {
             do {
                 String processName = Native.toString(pe32.szExeFile);
                 if(processNames.contains(processName.toLowerCase(Locale.ENGLISH))) {
-                    javaPids.add(pe32.th32ProcessID.toString());
+                    pidList.add(pe32.th32ProcessID.toString());
                 }
             } while (Kernel32.INSTANCE.Process32Next(hSnapshot, pe32));
         }
 
         Kernel32.INSTANCE.CloseHandle(hSnapshot);
-        return javaPids;
+        return pidList;
     }
 
 
     @Override
     public ArrayList<Path> getPidPaths(ArrayList<String> pids) {
-        ArrayList<Path> paths = new ArrayList<>();
+        ArrayList<Path> pathList = new ArrayList<>();
 
         for(String pid : pids) {
             WinNT.HANDLE hProcess = Kernel32.INSTANCE.OpenProcess(WinNT.PROCESS_QUERY_INFORMATION | WinNT.PROCESS_VM_READ, false, Integer.parseInt(pid));
@@ -104,11 +107,11 @@ public class WindowsAppLocator extends AppLocator{
             }
 
             Kernel32.INSTANCE.CloseHandle(hProcess);
-            paths.add(Paths.get(Native.WCHAR_SIZE == 1 ?
+            pathList.add(Paths.get(Native.WCHAR_SIZE == 1 ?
                                         buffer.getString(0) :
                                         buffer.getWideString(0)));
         }
-        return paths;
+        return pathList;
     }
 
     /**

--- a/src/qz/installer/certificate/firefox/locator/WindowsAppLocator.java
+++ b/src/qz/installer/certificate/firefox/locator/WindowsAppLocator.java
@@ -10,31 +10,28 @@
 
 package qz.installer.certificate.firefox.locator;
 
-import org.apache.commons.lang3.StringUtils;
+import com.sun.jna.Memory;
+import com.sun.jna.Native;
+import com.sun.jna.Pointer;
+import com.sun.jna.platform.win32.Kernel32;
+import com.sun.jna.platform.win32.Psapi;
+import com.sun.jna.platform.win32.Tlhelp32;
+import com.sun.jna.platform.win32.WinNT;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 import qz.installer.certificate.firefox.locator.AppAlias.Alias;
-import qz.utils.ShellUtilities;
-import qz.utils.SystemUtilities;
 import qz.utils.WindowsUtilities;
 
 
-import java.io.File;
-import java.io.IOException;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.ArrayList;
+import java.util.Locale;
 
 import static com.sun.jna.platform.win32.WinReg.HKEY_LOCAL_MACHINE;
 
 public class WindowsAppLocator extends AppLocator{
     protected static final Logger log = LogManager.getLogger(MacAppLocator.class);
-
-    private static final String[] WIN32_PID_QUERY = {"wmic.exe", "process", "where", null, "get", "processid"};
-    private static final int WIN32_PID_QUERY_INPUT_INDEX = 3;
-
-    private static final String[] WIN32_PATH_QUERY = {"wmic.exe", "process", "where", null, "get", "ExecutablePath"};
-    private static final int WIN32_PATH_QUERY_INPUT_INDEX = 3;
 
     private static String REG_TEMPLATE = "Software\\%s%s\\%s%s";
 
@@ -61,48 +58,57 @@ public class WindowsAppLocator extends AppLocator{
 
     @Override
     public ArrayList<String> getPids(ArrayList<String> processNames) {
-        ArrayList<String> pidList = new ArrayList<>();
+        ArrayList<String> javaPids = new ArrayList<>();
+        Tlhelp32.PROCESSENTRY32 pe32 = new Tlhelp32.PROCESSENTRY32();
+        pe32.dwSize = new WinNT.DWORD(pe32.size());
 
-        if (processNames.isEmpty()) return pidList;
-
-        WIN32_PID_QUERY[WIN32_PID_QUERY_INPUT_INDEX] = "(Name='" + String.join("' OR Name='", processNames) + "')";
-        String[] response = ShellUtilities.executeRaw(WIN32_PID_QUERY).split("[\\r\\n]+");
-
-        // Add all found pids
-        for(String line : response) {
-            String pid = line.trim();
-            if(StringUtils.isNumeric(pid.trim())) {
-                pidList.add(pid);
-            }
+        // Fetch a snapshot of all processes
+        WinNT.HANDLE hSnapshot = Kernel32.INSTANCE.CreateToolhelp32Snapshot(Tlhelp32.TH32CS_SNAPPROCESS, new WinNT.DWORD(0));
+        if (hSnapshot.equals(WinNT.INVALID_HANDLE_VALUE)) {
+            log.warn("Process snapshot has invalid handle");
+            return javaPids;
         }
 
-        if(WindowsUtilities.isWindowsXP()) {
-            // Cleanup XP crumbs per https://stackoverflow.com/q/12391655/3196753
-            File f = new File("TempWmicBatchFile.bat");
-            if(f.exists()) {
-                f.deleteOnExit();
-            }
+        if (Kernel32.INSTANCE.Process32First(hSnapshot, pe32)) {
+            do {
+                String processName = Native.toString(pe32.szExeFile);
+                if(processNames.contains(processName.toLowerCase(Locale.ENGLISH))) {
+                    javaPids.add(pe32.th32ProcessID.toString());
+                }
+            } while (Kernel32.INSTANCE.Process32Next(hSnapshot, pe32));
         }
 
-        return pidList;
+        Kernel32.INSTANCE.CloseHandle(hSnapshot);
+        return javaPids;
     }
+
 
     @Override
     public ArrayList<Path> getPidPaths(ArrayList<String> pids) {
-        ArrayList<Path> pathList = new ArrayList<>();
+        ArrayList<Path> paths = new ArrayList<>();
 
         for(String pid : pids) {
-            WIN32_PATH_QUERY[WIN32_PATH_QUERY_INPUT_INDEX] = "ProcessId=" + pid;
-            String[] response = ShellUtilities.executeRaw(WIN32_PATH_QUERY).split("\\s*\\r?\\n");
-            if (response.length > 1) {
-                try {
-                    pathList.add(Paths.get(response[1]).toRealPath());
-                } catch(IOException e) {
-                    log.warn("Could not locate process " + pid);
-                }
+            WinNT.HANDLE hProcess = Kernel32.INSTANCE.OpenProcess(WinNT.PROCESS_QUERY_INFORMATION | WinNT.PROCESS_VM_READ, false, Integer.parseInt(pid));
+            if (hProcess == null) {
+                log.warn("Handle for PID {} is missing, skipping.", pid);
+                continue;
             }
+
+            int bufferSize = WinNT.MAX_PATH;
+            Pointer buffer = new Memory(bufferSize * Native.WCHAR_SIZE);
+
+            if (Psapi.INSTANCE.GetModuleFileNameEx(hProcess, null, buffer, bufferSize) == 0) {
+                log.warn("Full path to PID {} is empty, skipping.", pid);
+                Kernel32.INSTANCE.CloseHandle(hProcess);
+                continue;
+            }
+
+            Kernel32.INSTANCE.CloseHandle(hProcess);
+            paths.add(Paths.get(Native.WCHAR_SIZE == 1 ?
+                                        buffer.getString(0) :
+                                        buffer.getWideString(0)));
         }
-        return pathList;
+        return paths;
     }
 
     /**

--- a/src/qz/utils/ShellUtilities.java
+++ b/src/qz/utils/ShellUtilities.java
@@ -182,10 +182,6 @@ public class ShellUtilities {
         InputStreamReader in = null;
         try {
             Process p = Runtime.getRuntime().exec(commandArray, envp);
-            if(SystemUtilities.isWindows() && commandArray.length > 0 && commandArray[0].startsWith("wmic")) {
-                // Fix deadlock on old Windows versions https://stackoverflow.com/a/13367685/3196753
-                p.getOutputStream().close();
-            }
             in = new InputStreamReader(p.getInputStream(), Charsets.UTF_8);
             StringBuilder out = new StringBuilder();
             int c;

